add job.sumbit service, flux-submit(1) command, and global sequences for jobids

grondo opened the pull request:
This PR is still a bit of a work in progress, but I wanted to open it and get early comments.

This PR reworks the `job` module a bit and adds a simple `job.submit` service. The `job.submit` service simply creates a new job entry and sets the `submitted` state, iff the `sched` module is loaded -- returning ENOSYS otherwise.

New bindings are included in the `wreck.lua` module for `wreck:createjob()` and `wreck:submit()` to abstract common code, and a `flux-submit` command has been added directly to core. The `submit` command in flux-sched will still work, but can be removed once tests are updated to use the builtin submit command.

This PR also adds a simple named sequence service directly to the broker, and uses a `"lwj"` sequence to allocate jobids instead of using the `lwj.next-id` kvs entry. A new `flux wreck last-jobid` command is added to return the current value of the `lwj` sequence number.

Migrating to a global sequence id also allows the job module to now be loaded on every rank.

Still TODO:
 - Is it ok for simple sequence service to be implemented directly in the broker, or should this be another module?
 - If we want `job.submit` to be functional without sched loaded, and instead move jobs directly to `runrequest`, then a task layout service will have to be added to assign the jobs tasks to ranks. This could be something simple, but I wonder if it is necessary if flux-core will eventually have its own simple sched?
 - man page for flux-submit(1) 
 - flux-sched tests appear to pass against this branch, however, there is still a watch used in `simulator/scheduler.c` against `lwj.next-id` that probably needs to be addressed. Also, I left in the code that implements the `race_workaround` for the simulator -- but I wonder if this is still needed?

garlick commented:
> Is it ok for simple sequence service to be implemented directly in the broker, or should this be another module?

IMHO this is is fine.

I just walked through that part of the code and looks great.  Couple of minor comments:

`zhash_insert()` failure case in `seq_create()` should probably be an assertion, or just use `zhash_update()`

I think at one point you had a single (or a couple?) methods instead of the six here (create, destroy, set, next, fetch, fetch-and-add) while providing the same functionality.  Did that lead to confusing code?  It would seem desirable to minimize the API footprint and number of different messages that would (eventually) go in an RFC.

Speaking of API, should this service be exposed in the C API?

grondo commented:
Thanks for the comments!

> zhash_insert() failure case in seq_create() should probably be an assertion, or just use zhash_update()

Ok, good point. I think the assertion is better here since we don't want to recreate and existing sequence.

>  think at one point you had a single (or a couple?) methods instead of the six here (create, destroy, set, next, fetch, fetch-and-add) while providing the same functionality. Did that lead to confusing code? 

It lead to a bit of a confusing interface. I guess I found that either you create the functionality in the method/topic names, or you push the complexity into the message contents. When the functionality is spelled out in the topic string, I found it much more user friendly than to push a bunch of complexity into the message itself, which led to several oops along the way. (E.g. if you want to increment and fetch, but you typo `increment` you will still get back a reply -- and accidentally reuse a jobid. If you typo the topic string, you'll get back ENOSYS)

Now whether these methods are even necessary at all is a different question and largely up for debate in my mind...

> Speaking of API, should this service be exposed in the C API?

Hm, I thought about that, but I think I have to be convinced it is necessary and not just gratuitous. Since these are "simple" send/recv API a C binding around `flux_rpc` doesn't seem to add much, and would be one more place to update code when there are changes. (it also doesn't handle the non-blocking case, where you would want to use non-blocking flux_rpc or send/recv). What is the benefit of C api in this case? (Honestly asking here) I'd be perfectly willing to add it if so.

lipari commented:
Just checking in to say I looked over the PR and it looks great!  I liked the way you minimized the complexity of the new flux-submit by moving all of the argument processing into common code.  W/r/t your question on whether to add a task layout service, I don't see that that is necessary at this time.  It can be deferred to the effort to create core's simple sched.

grondo commented:
Thanks for comments @lipari!

grondo commented:
Oh, one other question that came up during this work is that currently I preserve the job state ordering from before during job submit -- i.e. for `job.submit` the new lwj still goes first into `reserved` state then almost immediately into `submitted` (when sched is available).

I wasn't sure if anything would break if jobs skipped the `reserved` state, but we could look into having different "initial" states for jobs to reduce the number of events emitted per job.

lipari commented:
The reserved -> submitted transition was used to signify a candidate for scheduling.  The reserved state is not needed by sched and is actually irrelevant at this point.  If it went away, we would make a minor change sched's state machine to support a null -> submitted state transition and rip out the intermediate reserved state transition.

dongahn commented:
I agree with @lipari. I attached the current FSM that the sched uses. This should change once we hook dynamic scheduling into real runtime actions of course.

Dong
----------------------------------------------------------
Dong H. Ahn
925-422-1939
Development Environment  Group
Livermore Computing Division
Lawrence Livermore National Laboratory

From: Don Lipari <notifications@github.com<mailto:notifications@github.com>>
Reply-To: flux-framework/flux-core <reply@reply.github.com<mailto:reply@reply.github.com>>
Date: Sunday, November 15, 2015 at 2:29 PM
To: flux-framework/flux-core <flux-core@noreply.github.com<mailto:flux-core@noreply.github.com>>
Subject: Re: [flux-core] add job.sumbit service, flux-submit(1) command, and global sequences for jobids (#480)


The reserved -> submitted transition was used to signify a candidate for scheduling. The reserved state is not needed by sched and is actually irrelevant at this point. If it went away, we would make a minor change sched's state machine to support a null -> submitted state transition and rip out the intermediate reserved state transition.

—
Reply to this email directly or view it on GitHub<https://github.com/flux-framework/flux-core/pull/480#issuecomment-156850576>.

grondo commented:
>  instead of the six here (create, destroy, set, next, fetch, fetch-and-add) 

Ok, thinking about this a bit more, I could roll fetch and fetch-and-add together if the `increment` parameter is replaced with `preincrement` and `postincrement`. Also, `next` is the same as `fetch` with `increment = 1` and `create = true`, so this is unnecessary, though I found it a much nicer interface for simple sequence usage as by the `job` module. `create` could also be replaced by `fetch` with `create = true` -- if the message results a creation then `created = true` could be returned to the caller for callers that are interested in whether their own message resulted in the create.  This would leave us with something like:

 * **fetch** 
```
{ name: STRING,
  create: BOOLEAN,
  preincrement:  INT64,
  postincrement:  INT64,
}
```
returns
```
{ name: STRING,
  value: INT64,
  created: BOOLEAN
}
```
`created = true` is set on return value if the original message resulted in creation of a new sequence. New sequence numbers will start at 0, and have preincrement applied before the response, and postincrement after. A `fetch` message with preincrement and postincrement set to zero would return the current value of the sequence.

 * **destroy**
```
 { name = STRING }
```

If the members of the `fetch` method are required instead of optional, then this satisfies my complaint about typos in json keys, with a drawback of larger messages for the common case.

@garlick, would this be more acceptable? With this simplified message interface, the argument for a C API might be a bit stronger.

garlick commented:
That sounds a little better to me but I would defer to your judgement, since you've actually used the service.

Regarding async interface in C, I like the way `cas_load()` worked out in PR #471, where I wrapped `flux_rpc()` and `flux_rpc_get()`.  That allows the interface to be used asynchronously with `flux_rpc_then()` on the rpc returned by `cas_load()`, or synchronously by calling `cas_load_get()` immediately.

grondo commented:
> That sounds a little better to me

What would make it a lot better? I guess I don't want to add this service to the broker if the design is felt to be too unwieldy. Maybe it should either be its own module, so it could be more easily replaced int the future, or 2. stripped down to just an increment service that would return a sequence number always incremented by 1 (which is really the only use case currently) I was attempting to be forward thinking (the dumb increment service just seemed a bit naive), but perhaps it isn't time for that.

> That allows the interface to be used asynchronously with flux_rpc_then() on the rpc returned by cas_load(), or synchronously by calling cas_load_get() immediately.

I guess my point here is one could get the same benefit using `flux_rpc` directly. So I guess the main benefit of C wrapper is hidden marshal/unmarshall of the JSON message? I was wondering if that was worth the effort for this set of functions... It sounds like it is (if we keep this service) so I can work on that.

grondo commented:
Ok, I've pushed a new version with the changes to `cmb.seq` discussed above. The C api and other items are still todo

garlick commented:
`sequence_hash_create()` should handle a NULL return from `zhash_new()`.   The sequence stuff looks good to me and I think a C API can wait until there is a demonstrated need.

grondo commented:
Great, thanks for the review @garlick!

trws commented:
This looks pretty great to me @grondo.  I do have a couple of questions, just to make sure I'm interpreting this correctly.

1) The job service runs independently on all broker ranks, so only when it actually submits does it hit the scheduler, right?
2) The sequence service appears to run only on rank 0, but uses an un-targeted request.  If we decided to change to a decentralized consensus model later could that request be intercepted by the local broker or will this need to remain centralized?
3) Detecting a failed or invalid submission seems to be through looking for ENOSYS as the return or error code, does the lua interface make that visible? Is there a scheme to determine the type of failure?

As to a C interface, I don't really see a reason right now.  As the jobspec stuff comes together there should be a C interface to ease creating the appropriate request object, but rpc seems sufficient to actually manipulate the messages.

garlick commented:
> 2) The sequence service appears to run only on rank 0, but uses an un-targeted request. If we decided to change to a decentralized consensus model later could that request be intercepted by the local broker or will this need to remain centralized?

If the sequence service appeared on other ranks (for example the "top" 5-7 ranks in the tree that had formed a raft cluster), I believe this would just work without modifying users.

grondo commented:
> The job service runs independently on all broker ranks, so only when it actually submits does it hit the scheduler, right?

Yes, currently `job.submit` runs independently on all ranks, and the job module on each individual rank would issue the "wrexec.submitted" event to notify scheduler of work.

> Detecting a failed or invalid submission seems to be through looking for ENOSYS as the return or error code, does the lua interface make that visible? Is there a scheme to determine the type of failure?

If sched is not loaded then a response message will be issued with `errnum == ENOSYS`. If the message itself is malformed then `errnum == EPROTO`. These error numbers can be determined from either the C or Lua APIs through standard means. In Lua it is something like:
```lua
local resp, err = f:rpc ("job.submit", request)
if not resp then error (err) end
```
Though currently `err` is a string which might not be ideal. We could also pass the error number (as luaposix does) and change that idiom to:
```lua
local resp, err, errnum = f:rpc ("job.submit", request)
```
In a Lua msghandler, the errnum is available as a member of the returned `zmsg` object (still using old zmsg based message types in Lua bindings)

TBD is to add the "filter" service to job.submit and job.create services. In these cases, it might be more user friendly to extend the response message to contain an error flag and "reason" or similar string. This work is still in design stages and pretty much open to do whatever we want. We could replace the `errnum` based return messages described above with this scheme as well if that would be more tractable. (e.g. missing sched would return  `{ error = true, reason = "no sched module found"}`).

grondo commented:
Ok, I've added `flux-submit(1)` manpage and updated the `flux-wreck(1)` manpage with a blurb on `last-jobid` and addressed @garlick's comment on `sequence_hash_create`.

`make check` succeeds in flux-sched master against this branch,  but `simultator/scheduler.c` still does add a watch on `lwj.next-id`, which I think will need to be addressed (Haven't investigated further yet).

After this branch is merged, I'll submit a PR to flux-sched removing the version of `flux-submit` in that tree, and with some other tests-related cleanup.

garlick commented:
Looks great to me.   I'm happy to hit the big green button once someone from the sched team ACKs whether it's OK given the simulator's use of `lwj.next-id`.  Nice work @grondo.

dongahn commented:
@grondo I will talk with Stephen about the simulator/scheduler.c when I come across him again (I was with him about 10 mins ago.) 

I don't think that's used any more...but I need to doublecheck. 

 In any case, now that we almost completed our CCGRID submission we should spend sime time to clean up some code before making a big move.

garlick commented:
So I think we're just waiting to get an ACK from @SteVwonder that this is OK to merge?

dongahn commented:
Yes he plans to look at it either tonight or tomorrow.

BTW our SC15 Flux poster has been selected to advance to the next round toward a Best poster award so he's busy prepping for the oral presentation (this afternoon) as we speak. Only 5 posters moved on; this is awesome!

Best
Dong
________________________________________
From: Jim Garlick [notifications@github.com]
Sent: Wednesday, November 18, 2015 8:33 AM
To: flux-framework/flux-core
Cc: Ahn, Dong H.
Subject: Re: [flux-core] add job.sumbit service, flux-submit(1) command, and global sequences for jobids (#480)

So I think we're just waiting to get an ACK from @SteVwonder<https://github.com/SteVwonder> that this is OK to merge?

�
Reply to this email directly or view it on GitHub<https://github.com/flux-framework/flux-core/pull/480#issuecomment-157769851>.

grondo commented:
Ok, with flux-framework/flux-sched#95 merged, this should be safe to merge as well.
